File: SPCMeasurement/SPCMeasurement/clsItemCheck.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Data;
using System.Data.SqlClient;
using C1.Win.C1List;

namespace SPCMeasurement
{
    class clsItemCheck
    {
        public string ItemCheckCode { get; set; }
        public string ItemCheck { get; set; }
        public string PrevItemCheck { get; set; }
        public string PrevValue { get; set; }
    }

    class clsItemCheckDB
    {
        private static clsConfig cfg = new clsConfig();
        private static clsConfigData cfd = cfg.uf_ReadConfig();
        private static string constr = cfd.ConnectionString;

        public static void FillCombo(C1.Win.C1List.C1Combo cbo, string FactoryCode, string LineCode, string ItemTypeCode)
        {

            cbo.DataMode = DataModeEnum.Normal;
            using (SqlConnection cn = new SqlConnection(constr))
            {
                cn.Open();
                string q;
                q = "select I.ItemCheckCode, I.ItemCheckCode + ' - ' + I.ItemCheck ItemCheck, " +
                    "isnull(Measure2Cls, '0') Measure2Cls, " +
                    "isnull(T.PrevItemCheck, '') PrevItemCheck, " + 
                    "case when isnull(T.PrevItemCheck, '') = '' then 0 " +
                    "when isnull(T.PrevValue, '') = '1' then 1 " +
                    "else 2 end PrevValue " +
                    "from spc_ItemCheckMaster I inner join spc_ItemCheckByType T on I.ItemCheckCode = T.ItemCheckCode " +
                    "where T.ItemCheckCode is not Null ";
                if (FactoryCode != "") {
                    q = q + "and T.FactoryCode = @FactoryCode ";
                } 
                if (ItemTypeCode != "")
                {
                    q = q + "and ItemTypeCode = @ItemTypeCode ";
                }
                if (LineCode != "") { }
                {
                    q = q + "and LineCode = @LineCode ";
                }
                q = q + "and T.ActiveStatus = '1' ";
                SqlCommand cmd = new SqlCommand(q, cn);
                cmd.Parameters.AddWithValue("FactoryCode", FactoryCode);
                cmd.Parameters.AddWithValue("ItemTypeCode", ItemTypeCode);
                cmd.Parameters.AddWithValue("LineCode", LineCode);
                SqlDataAdapter da = new SqlDataAdapter(cmd);
                DataTable dt = new DataTable();
                da.Fill(dt);
                cbo.DataSource = dt;
                int n = dt.Rows.Count;
                cbo.ColumnHeaders = false;
                cbo.DataMode = DataModeEnum.Normal;
                cbo.ItemHeight = 18;
                cbo.MaxDropDownItems = 10;
                cbo.Splits[0].DisplayColumns[0].Visible = false;
                cbo.Splits[0].DisplayColumns[2].Visible = false;
                cbo.Splits[0].DisplayColumns[3].Visible = false;
                cbo.Splits[0].DisplayColumns[4].Visible = false;
                cbo.ExtendRightColumn = true;                
                cbo.DropdownPosition = C1.Win.C1List.DropdownPositionEnum.LeftDown;
                cbo.ValueMember = "ItemCheckCode";
                cbo.DisplayMember = "ItemCheck";
                cbo.LimitToList = true;
            }
        }

        public static List<clsItemCheck> GetPrevItemChek(string FactoryCode, string ItemTypeCode, string Line, string ItemCheckCode, string ProdDate, string Shift, int Sequence, string UserID)
        {
            using (SqlConnection con = new SqlConnection(constr))
            {
                con.Open();
                SqlCommand cmd = new SqlCommand("sp_SPC_GetPrevItemCheck", con);
                cmd.Parameters.AddWithValue("FactoryCode", FactoryCode);
                cmd.Parameters.AddWithValue("Line", Line);
                cmd.Parameters.AddWithValue("ItemCheckCode", ItemCheckCode);
                cmd.Parameters.AddWithValue("ProdDate", ProdDate);
                cmd.Parameters.AddWithValue("ShiftCode", Shift);
                cmd.Parameters.AddWithValue("SequenceNo", Sequence);
                cmd.Parameters.AddWithValue("UserID", UserID);
                cmd.CommandType = CommandType.StoredProcedure;
                SqlDataReader rd = cmd.ExecuteReader();
                List<clsItemCheck> result = new List<clsItemCheck>();
                while (rd.Read())
                {
                    clsItemCheck item = new clsItemCheck();
                    item.ItemCheckCode = rd["ItemCheckCode"].ToString();
                    item.ItemCheck = rd["ItemCheck"].ToString();
                    result.Add(item);
                }
                rd.Close();
                return result;
            }
        }

        public static clsItemCheck GetItemChek(string ItemCheckCode)
        {
            using (SqlConnection con = new SqlConnection(constr))
            {
                con.Open();
                string q = "select T.*, I.ItemCheck from spc_ItemCheckByType T inner join spc_ItemCheckMaster I on T.ItemCheckCode = I.ItemCheckCode where T.ItemCheckCode = @ItemCheckCode";
                SqlCommand cmd = new SqlCommand(q, con);
                cmd.Parameters.AddWithValue("ItemCheckCode", ItemCheckCode);
                SqlDataAdapter da = new SqlDataAdapter(cmd);
                DataTable dt = new DataTable();
                da.Fill(dt);
                if(dt.Rows.Count > 0)
                {
                    clsItemCheck item = new clsItemCheck();
                    item.ItemCheckCode = dt.Rows[0]["ItemCheckCode"].ToString();
                    item.PrevItemCheck = dt.Rows[0]["PrevItemCheck"] + "";
                    item.PrevValue = dt.Rows[0]["PrevValue"] + "";
                    return item;
                } else
                {
                    return null;
                }
            }
        }
    }
}

File: PECGI_SPC/Class/clsItemCheck.vb
Imports System.Data.SqlClient

Public Class clsItemCheck
    Public Property ItemCheckCode As String
    Public Property ItemCheck As String
    Public Property Measure2Cls As String
End Class

Public Class clsItemCheckDB
    Public Shared Function GetData(ItemCheckCode As String) As clsItemCheck
        Using Cn As New SqlConnection(Sconn.Stringkoneksi)
            Cn.Open()
            Dim q As String = "Select * from spc_ItemCheckMaster where ItemCheckCode = @ItemCheckCode"
            Dim cmd As New SqlCommand(q, Cn)
            cmd.Parameters.AddWithValue("ItemCheckCode", ItemCheckCode)
            Dim da As New SqlDataAdapter(cmd)
            Dim dt As New DataTable
            da.Fill(dt)
            If dt.Rows.Count = 0 Then
                Return Nothing
            Else
                Dim Item As New clsItemCheck
                Item.ItemCheckCode = dt.Rows(0)("ItemCheckCode") & ""
                Item.ItemCheck = dt.Rows(0)("ItemCheck") & ""
                Item.Measure2Cls = dt.Rows(0)("Measure2Cls") & ""
                Return Item
            End If
        End Using
    End Function

    Public Shared Function GetList(Optional FactoryCode As String = "", Optional ItemTypeCode As String = "", Optional LineCode As String = "", Optional ShowAll As Boolean = False) As List(Of clsItemCheck)
        Using Cn As New SqlConnection(Sconn.Stringkoneksi)
            Cn.Open()
            Dim q As String = ""
            If ShowAll Then
                q = "select 'ALL' ItemCheckCode, 'ALL' ItemCheck, 0 seq union "
            End If
            q = q & "select I.ItemCheckCode, I.ItemCheckCode + ' - ' + I.ItemCheck ItemCheck, 1 seq " & vbCrLf &
                "from spc_ItemCheckMaster I inner join spc_ItemCheckByType T on I.ItemCheckCode = T.ItemCheckCode "
            q = q & "where T.ItemCheckCode is not Null "
            If FactoryCode <> "" Then
                q = q & "and T.FactoryCode = @FactoryCode " & vbCrLf
            End If
            If ItemTypeCode <> "" Then
                q = q & "and ItemTypeCode = @ItemTypeCode " & vbCrLf
            End If
            If LineCode <> "" Then
                q = q & "and LineCode = @LineCode "
            End If
            q = q & "and T.ActiveStatus = '1' "
            q = q & "order by seq, ItemCheckCode "
            Dim cmd As New SqlCommand(q, Cn)
            cmd.Parameters.AddWithValue("FactoryCode", FactoryCode)
            cmd.Parameters.AddWithValue("ItemTypeCode", ItemTypeCode)
            cmd.Parameters.AddWithValue("LineCode", LineCode)
            Dim rd As SqlDataReader = cmd.ExecuteReader
            Dim ItemCheckList As New List(Of clsItemCheck)
            Do While rd.Read
                Dim ItemCheck As New clsItemCheck
                ItemCheck.ItemCheckCode = rd("ItemCheckCode") & ""
                ItemCheck.ItemCheck = rd("ItemCheck") & ""
                ItemCheckList.Add(ItemCheck)
            Loop
            rd.Close()
            Return ItemCheckList
        End Using
    End Function
End Class
